File: mobile/test/stage_permissions_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:kaede_mobile/src/core/refs.dart';
import 'package:kaede_mobile/src/domain/models.dart';
import 'package:kaede_mobile/src/domain/stage_permissions.dart';
import 'package:kaede_mobile/src/protocol/generated.dart';

void main() {
  KaedeChannel channel(ChannelType type, int permissions) => KaedeChannel(
        ref: EntityRef.parse('13@chat.example'),
        guildRef: EntityRef.parse('1@chat.example'),
        type: type,
        position: 0,
        permissions: BigInt.from(permissions),
      );

  test('Stage lifecycle requires the complete moderator trio', () {
    expect(
      canManageStageChannel(
        channel(ChannelType.stage, stageModeratorPermissions),
      ),
      isTrue,
    );
    expect(
      canManageStageChannel(
        channel(
          ChannelType.stage,
          Permission.manageChannels | Permission.muteMembers,
        ),
      ),
      isFalse,
    );
    expect(
      canManageStageChannel(
        channel(ChannelType.stage, Permission.administrator),
      ),
      isTrue,
    );
  });

  test('scheduled-event channels use Stage and voice-specific access', () {
    expect(
      canCreateScheduledEventInChannel(
        channel(
          ChannelType.stage,
          Permission.createEvents | stageModeratorPermissions,
        ),
      ),
      isTrue,
    );
    expect(
      canCreateScheduledEventInChannel(
        channel(ChannelType.stage, Permission.createEvents),
      ),
      isFalse,
    );
    expect(
      canCreateScheduledEventInChannel(
        channel(
          ChannelType.voice,
          Permission.createEvents | Permission.viewChannel | Permission.connect,
        ),
      ),
      isTrue,
    );
  });

  test('event management keeps ownership rules plus Stage moderation', () {
    expect(
      canManageScheduledEventInChannel(
        channel(
          ChannelType.stage,
          Permission.createEvents | stageModeratorPermissions,
        ),
        ownEvent: true,
      ),
      isTrue,
    );
    expect(
      canManageScheduledEventInChannel(
        channel(
          ChannelType.stage,
          Permission.manageEvents | stageModeratorPermissions,
        ),
        ownEvent: false,
      ),
      isTrue,
    );
    expect(
      canManageScheduledEventInChannel(
        channel(ChannelType.stage, Permission.manageEvents),
        ownEvent: false,
      ),
      isFalse,
    );
  });

  test('server-deafen is hidden in Stage but retained in voice', () {
    expect(
      canServerDeafenInChannel(
        channel(ChannelType.stage, Permission.deafenMembers),
      ),
      isFalse,
    );
    expect(
      canServerDeafenInChannel(
        channel(ChannelType.voice, Permission.deafenMembers),
      ),
      isTrue,
    );
  });
}

File: frontend/src/lib/voice/stage-permissions.test.ts
import { describe, expect, it } from 'vitest';
import { Permission } from '$lib/generated/permissions';
import {
  STAGE_MODERATOR_PERMISSIONS,
  canCreateScheduledEventInChannel,
  canManageScheduledEventInChannel,
  canManageStageChannel,
  canServerDeafenInChannel
} from './stage-permissions';

const channel = (type: number, permissions: bigint) => ({
  type,
  permissions: permissions.toString()
});

describe('Stage permission predicates', () => {
  it('requires the complete moderator trio for Stage lifecycle controls', () => {
    expect(canManageStageChannel(channel(13, STAGE_MODERATOR_PERMISSIONS))).toBe(true);
    expect(
      canManageStageChannel(channel(13, Permission.MANAGE_CHANNELS | Permission.MUTE_MEMBERS))
    ).toBe(false);
    expect(canManageStageChannel(channel(13, Permission.ADMINISTRATOR))).toBe(true);
    expect(canManageStageChannel(channel(2, STAGE_MODERATOR_PERMISSIONS))).toBe(false);
  });

  it('uses Stage and voice-specific scheduled-event access', () => {
    expect(
      canCreateScheduledEventInChannel(
        channel(13, Permission.CREATE_EVENTS | STAGE_MODERATOR_PERMISSIONS)
      )
    ).toBe(true);
    expect(canCreateScheduledEventInChannel(channel(13, Permission.CREATE_EVENTS))).toBe(false);
    expect(
      canCreateScheduledEventInChannel(
        channel(2, Permission.CREATE_EVENTS | Permission.VIEW_CHANNEL | Permission.CONNECT)
      )
    ).toBe(true);
    expect(
      canCreateScheduledEventInChannel(channel(2, Permission.CREATE_EVENTS | Permission.CONNECT))
    ).toBe(false);
  });

  it('preserves own-vs-other management while requiring Stage moderation', () => {
    expect(
      canManageScheduledEventInChannel(
        channel(13, Permission.CREATE_EVENTS | STAGE_MODERATOR_PERMISSIONS),
        true
      )
    ).toBe(true);
    expect(
      canManageScheduledEventInChannel(
        channel(13, Permission.MANAGE_EVENTS | STAGE_MODERATOR_PERMISSIONS),
        false
      )
    ).toBe(true);
    expect(canManageScheduledEventInChannel(channel(13, Permission.MANAGE_EVENTS), false)).toBe(
      false
    );
    expect(
      canManageScheduledEventInChannel(
        channel(2, Permission.CREATE_EVENTS | Permission.VIEW_CHANNEL | Permission.CONNECT),
        true
      )
    ).toBe(true);
  });

  it('never exposes server-deafen in Stage channels', () => {
    expect(canServerDeafenInChannel(channel(13, Permission.DEAFEN_MEMBERS))).toBe(false);
    expect(canServerDeafenInChannel(channel(2, Permission.DEAFEN_MEMBERS))).toBe(true);
  });
});
